File: tests/fixtures/ruby/auth_service.rb
require 'bcrypt'
require_relative './user_repository'
require_relative './token_service'

# Authentication service fixture for coldstart parser tests.
module Auth
  class AuthService
    MAX_RETRIES = 3

    def initialize(user_repository, token_service)
      @user_repository = user_repository
      @token_service = token_service
    end

    def login(email, password)
      user = @user_repository.find_by_email(email)
      return AuthResult.failure('User not found') unless user

      valid = verify_password(password, user.password_hash)
      return AuthResult.failure('Invalid password') unless valid

      token = @token_service.sign(user.id)
      AuthResult.success(token)
    end

    def self.hash_password(password)
      BCrypt::Password.create(password)
    end

    private

    def verify_password(raw, hash)
      BCrypt::Password.new(hash) == raw
    end
  end
end

File: tests/fixtures/java/AuthService.java
package com.example.auth;

import com.example.user.UserRepository;
import com.example.token.TokenService;
import java.util.List;
import java.security.MessageDigest;

/**
 * Authentication service — used as a fixture for coldstart parser tests.
 */
public class AuthService implements Authenticatable {

    private static final String ALGORITHM = "SHA-256";
    private static final int MAX_RETRIES = 3;

    private final UserRepository userRepository;
    private final TokenService tokenService;

    public AuthService(UserRepository userRepository, TokenService tokenService) {
        this.userRepository = userRepository;
        this.tokenService = tokenService;
    }

    public AuthResult login(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail());
        if (user == null) {
            return AuthResult.failure("User not found");
        }
        boolean valid = verifyPassword(request.getPassword(), user.getPasswordHash());
        if (!valid) {
            return AuthResult.failure("Invalid password");
        }
        String token = tokenService.sign(user.getId());
        return AuthResult.success(token);
    }

    private boolean verifyPassword(String raw, String hash) {
        String hashed = hashPassword(raw);
        return hashed.equals(hash);
    }

    public static String hashPassword(String password) {
        try {
            MessageDigest digest = MessageDigest.getInstance(ALGORITHM);
            byte[] bytes = digest.digest(password.getBytes());
            return bytesToHex(bytes);
        } catch (Exception e) {
            throw new RuntimeException("Hash failed", e);
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}
